akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala

```diff
@@ -347,7 +347,10 @@ private class RestartSupervisor[O, T, Thr <: Throwable: ClassTag](initial: Behav
     }
 
     try {
-      val newBehavior = Behavior.validateAsInitial(Behavior.start(initial, ctx.asInstanceOf[TypedActorContext[T]]))
+      val startedInitial = Behavior.validateAsInitial(Behavior.start(initial, ctx.asInstanceOf[TypedActorContext[T]]))
+      // when withTimers has been installed it must always be installed so TimerInterceptor
+      // can discard TimerMsg from old generation/instance
+      val newBehavior = if (hasTimers(ctx)) Behaviors.withTimers[T](_ => startedInitial) else startedInitial
```

Comment:
Sounds like it could work. The only other option I can think of is to let the ActorAdapter know about timers and have some special handling rather than the current interceptor based impl.

Reply:
I think that can be a good alternative, I'll look into that.